cmd/clusterctl/client/repository/repository_github_test.go

```diff
@@ -337,15 +339,39 @@ func Test_gitHubRepository_getVersions(t *testing.T) {
 	client, mux, teardown := test.NewFakeGitHub()
 	defer teardown()
 
-	// setup an handler for returning 5 fake releases
+	// Building a fake GitHub releases slice.
+	releases := []string{}
+	for i := 0; i <= 5; i++ {
+		releases = append(releases, fmt.Sprintf(`{"id":%d, "tag_name": "v0.4.%d"}`, i+1, i)) // adding normal semver tags
+	}
+	releases = append(releases, `{"id":7, "tag_name": "v0.4.7-alpha"}`, `{"id":8, "tag_name": "v0.4.8-beta"}`, `{"id":9, "tag_name": "foo"}`, `{"id":10, "tag_name": "bar"}`) // adding pre-release and non-semver tags
+
+	pageSize := 3
+
 	mux.HandleFunc("/repos/o/r1/releases", func(w http.ResponseWriter, r *http.Request) {
 		testMethod(t, r, "GET")
+		page := r.URL.Query().Get("page")
+		var pageInt int
+		var err error
+		if page != "" {
+			pageInt, err = strconv.Atoi(page)
+			if err != nil {
+				t.Fatalf("unexpected error converting string to int: %v", err)
+			}
+		} else {
+			pageInt = 1
+		}
+
+		numPages := len(releases) / pageSize
+		if len(releases)%pageSize > 0 {
+			numPages++
+		}
+		if pageInt != numPages {
+			// Page 1 to (numPages-1) of releases
+			w.Header().Set("Link", fmt.Sprintf(`<https://api.github.com/repositories/12345/releases?page=%d>; rel="next"`, pageInt+1))
+		}
 		fmt.Fprint(w, `[`)
-		fmt.Fprint(w, `{"id":1, "tag_name": "v0.4.0"},`)
-		fmt.Fprint(w, `{"id":2, "tag_name": "v0.4.1"},`)
-		fmt.Fprint(w, `{"id":3, "tag_name": "v0.4.2"},`)
-		fmt.Fprint(w, `{"id":4, "tag_name": "v0.4.3-alpha"},`) // prerelease
-		fmt.Fprint(w, `{"id":5, "tag_name": "foo"}`)           // no semantic version tag
+		fmt.Fprint(w, returnPageResults(releases, pageSize, pageInt))
 		fmt.Fprint(w, `]`)
 	})
```

Comment:
How about this:
```go
mux.HandleFunc("/repos/o/r1/releases", func(w http.ResponseWriter, r *http.Request) {
	testMethod(t, r, "GET")
	page := r.URL.Query().Get("page")
	switch page {
	case "", "1":
		// Page 1
		w.Header().Set("Link", `<https://api.github.com/repositories/12345/releases?page=2>; rel="next"`) // Link to page 2
		fmt.Fprint(w, `[`)
		fmt.Fprint(w, `{"id":1, "tag_name": "v0.4.0"},`)
		fmt.Fprint(w, `{"id":2, "tag_name": "v0.4.1"}`)
		fmt.Fprint(w, `]`)
	case "2":
		// Page 2
		w.Header().Set("Link", `<https://api.github.com/repositories/12345/releases?page=3>; rel="next"`) // Link to page 3
		fmt.Fprint(w, `[`)
		fmt.Fprint(w, `{"id":3, "tag_name": "v0.4.2"},`)
		fmt.Fprint(w, `{"id":4, "tag_name": "v0.4.3-alpha"}`) // prerelease
		fmt.Fprint(w, `]`)
	case "3":
		// Page 3 (last page)
		fmt.Fprint(w, `[`)
		fmt.Fprint(w, `{"id":4, "tag_name": "v0.4.3-beta"},`) // prerelease
		fmt.Fprint(w, `{"id":5, "tag_name": "foo"}`)          // no semantic version tag
		fmt.Fprint(w, `]`)
	default:
		t.Fatalf("unexpected page requested")
	}
}
```
(also adjust the corresponding test's `want`. With this we will also not need `returnPageResults` and can be dropped)

IMO, it makes the tests more readable and easy to maintain.

Reply:
Thanks! Updated to use a static release listing and pagination. True, there's no need to complicate a unit testing with custom functions.